Layer with reusable weights and tables

liliatangxy opened the pull request:
Enables reuse of weights and table on top of the layer PR #27 and avoids recomputation for their setups. This also simplifies the handling of weights and tables for the user into providing a weights `HashMap` and configuration of tables by layer. 

Changes:
- Weights are stored in a `HashMap`, and are identified with a user-provided key provided to the `LayerConfig`
- `Graph` has a constructor that handles table creation based on a list of `LayerConfigs`
- Added `Setup` object and moved setup logic. Setups and weights/table `Data` objects are constructed once for each unique hash map entry
- Added `BasicBlockType`, `LayerType`, and for basic blocks `block_type()`, `weights_name()`
- Removed `setup` from `layer` and lower, changed signatures for `run`, etc., to not use `model` when possible

Vimalakirti committed:
add subgraph and layer features

Vimalakirti committed:
simplification: remove subgraph

Vimalakirti committed:
readability

Vimalakirti committed:
move nodes into layers

Vimalakirti committed:
oops

Vimalakirti committed:
debug and impl Hash

Vimalakirti committed:
add HashMap of basicblocks to load layer basicblocks

Vimalakirti committed:
fix cq name

Vimalakirti committed:
fix test

Vimalakirti committed:
add comment for fn hash

Vimalakirti committed:
edit comments for Hash

Vimalakirti committed:
edit comments for Hash

liliatangxy committed:
add reuse for weights and tables

liliatangxy committed:
cleanup

liliatangxy committed:
cleanup

liliatangxy committed:
remove comments

liliatangxy committed:
remove more dead code

ddkang commented:
Raise issues for the last 3 notes

ddkang commented on src/basic_block.rs:
```diff
@@ -89,24 +90,53 @@ impl DataEnc {
-  fn run(&self, _model: &ArrayD<Fr>, _inputs: &Vec<&ArrayD<Fr>>) -> Vec<ArrayD<Fr>> {
-    vec![]
+  fn weights_name(&self) -> String {
+    "".to_string()
```
Why don't we just error here?

ddkang commented on src/graph.rs:
```diff
@@ -9,14 +11,69 @@ pub struct Node {
+    let mut table_map: HashMap<String, ArrayD<Fr>> = HashMap::new();
+    for i in 1..basic_blocks.len() {
+      let table = match (basic_blocks[i - 1].block_type(), basic_blocks[i].block_type()) {
+        (BasicBlockType::ChangeSF, BasicBlockType::CQ2)
```
What's going on here?

Vimalakirti replied:
We need to initialize a CQ2 table for every nonlinear basicblock. With this realization, it's great that we no longer mix model weights and lookup tables together. But we could refactor codes (e.g., pack nonlinear basikblocktypes somewhere else) or add comments to make it more clear.

liliatangxy replied:
I simplified the logic here by checking whether a BB is a nonlinearity and erroring if it is not followed by CQ2. Right now table generation only uses CQ2 (combining range check with the nonlinearity), can introduce CQ logic in another PR.

ddkang commented on src/graph.rs:
```diff
@@ -9,14 +11,69 @@ pub struct Node {
+    }
+  }
+
+  pub fn run(&self, inputs: &Vec<&ArrayD<Fr>>) -> Vec<Vec<Vec<ArrayD<Fr>>>> {
```
What's going on with the vec vec vec stuff?

liliatangxy replied:
The indices are [layer no.][node id in that layer][the output index of that node].

@Vimalakirti Is there a reason why we need the output index?

Vimalakirti replied:
@liliatangxy we need that because some nodes may return two outputs (e.g., Split). But currently I think we don't have this kind of basicblock.

liliatangxy replied:
Ok, I think we can keep that in the code. I can add a comment to clarify this

ddkang commented on src/setup.rs:
```diff
@@ -0,0 +1,181 @@
+use crate::util::{self, convert_to_data};
```
Why is this a generic setup file?

ddkang commented:
Can we make a GitHub action to run tests when commits are pushed? Can raise an issue

Vimalakirti reviewed (approved):
Thanks! Except for readability, GitHub testing, and CustomLayer (we can delete it if unnecessary) issues, I have no other concerns. 
Btw, shall I close my previous PR since this PR is directly based on that one?

liliatangxy committed:
error handling and rearranging setup

liliatangxy committed:
fix error print

liliatangxy commented:
Cleaned up the graph constructor, added error checking for weights name and CQ2 following a nonlinearity, and moved setup back.
`name()` is actually used for the `Hash` trait, which is needed in the codebase so we don't need to remove it.

I think supporting `CustomLayer` would require us to put `Layer` inside the `LayerConfig`.

liliatangxy committed:
comment for graph run

Vimalakirti committed:
fix bug: duplicate basicblocks and wrong output node

Vimalakirti commented:
Hi, I found two bugs in `softmax` and `graph`:
- basicblock should not duplicate in graph initialization
- the output node of softmax is incorrect

I've pushed my patch for them, please check.

Vimalakirti committed:
oops

liliatangxy commented:
Thanks for catching them, the changes look good

liliatangxy committed:
debug trait and cleanup

liliatangxy committed:
remove customlayer

arifiorino commented on src/basic_block/cq2.rs:
```diff
@@ -15,20 +19,36 @@ use rand::{rngs::StdRng, SeedableRng};
+
+  fn name(&self) -> String {
+    // concat "CQ2" with self.name
+    format!("CQ2({})", self.name)
```
This is unnecessary because the `Debug` trait already includes `CQ2` and `self.name`.

liliatangxy replied:
`format` isn't sufficient for the basic blocks containing `HashTable` e.g., `CQ2`

arifiorino replied:
The `table_dict` parameter in `CQ2BasicBlock` doesn't contain anything when we're setting it up. It only gets populated during proof time. Therefore it shouldn't be an issue.

arifiorino commented on src/basic_block.rs:
```diff
@@ -80,24 +89,66 @@ impl DataEnc {
+
+pub trait BasicBlock: std::fmt::Debug {
+  fn block_type(&self) -> BasicBlockType {
+    BasicBlockType::Constant
```
Why would the default be `Constant`? I think there should be no default and every basic block should define this. But there is probably a more efficient way of doing this as well (maybe using `format!()` so we don't have to add code to every basic block.

arifiorino commented on src/basic_block.rs:
```diff
@@ -80,24 +89,66 @@ impl DataEnc {
-  fn setup(&self, _srs: &SRS, _model: &ArrayD<Data>) -> (Vec<G1Projective>, Vec<G2Projective>) {
-    (Vec::new(), Vec::new())
+
+  fn name(&self) -> String {
```
We should think about if we even need this method, because it might be better to just use `format!()` instead everywhere this method is used.

arifiorino commented on src/basic_block.rs:
```diff
@@ -80,24 +89,66 @@ impl DataEnc {
+    format!("{self:?}")
+  }
+
+  fn weights_name(&self) -> Result<String, String> {
```
I thought the weights would be part of the `name` itself.

arifiorino commented on src/basic_block.rs:
```diff
@@ -110,3 +161,22 @@ pub trait BasicBlock {
+  }
+}
+
+impl std::cmp::PartialEq for dyn BasicBlock {
```
This could also be done by adding `#[derive(Debug, PartialEq)]` to the basic blocks which might be more intuitive.

arifiorino commented on src/graph.rs:
```diff
@@ -1,66 +1,256 @@
+  pub table_map: HashMap<String, ArrayD<Fr>>,
 }
 
 impl Graph {
```
Shouldn't we just convert the layers into a `Graph`, and leave the `Graph` struct unchanged?

arifiorino commented on src/layer.rs:
```diff
@@ -0,0 +1,145 @@
+  pub weights_names: Vec<String>,
+}
+
+pub trait Layer {
```
So we have `run/prove/verify` defined for `BasicBlock`,`Layer`, and `Graph`? we don't need it defined for all these three.

arifiorino commented on src/layer/add.rs:
```diff
@@ -0,0 +1,20 @@
+use super::{Layer, LayerConfig};
```
This is too many lines (20) that we will have to write for every layer (there are 154 layers in the tract library)

arifiorino commented on src/graph.rs:
```diff
@@ -1,66 +1,256 @@
+  pub T_x_2: Vec<G2Affine>,
+}
+
+pub enum SetupType {
```
This seems a bit hardcoded to me. There might be a more flexible/simpler way to do this.

arifiorino commented on src/basic_block/cq.rs:
```diff
@@ -15,18 +18,34 @@ use rand::{rngs::StdRng, SeedableRng};
+    BasicBlockType::CQ
+  }
+
+  fn name(&self) -> String {
```
This is unnecessary because the `Debug` trait already includes `CQ` and `self.name`.

arifiorino commented on src/basic_block/cqlin.rs:
```diff
@@ -37,6 +55,9 @@ impl BasicBlock for CQLinBasicBlock {
     util::ifft_in_place::<G1Projective>(domain_m, &mut L_V_i_x);
+    let L_H_i_x = L_H_i_x.iter().map(|x| (*x).into()).collect();
+    let L_V_i_x_n = L_V_i_x_n.iter().map(|x| (*x).into()).collect();
+    let L_V_i_x = L_V_i_x.iter().map(|x| (*x).into()).collect();
```
Earlier, the setup functions returned `(Vec<G1Projective>,Vec<G2Projective>)` so we didn't have to write conversions like this into the setup functions.

arifiorino commented on src/basic_block/cq2.rs:
```diff
@@ -37,28 +57,41 @@ impl BasicBlock for CQ2BasicBlock {
-    setup.extend(L_i_0_x_1);
-    return (setup, setup2);
+    let L_i_x_1 = L_i_x_1.iter().map(|x| (*x).into()).collect();
+    let L_i_0_x_1 = L_i_0_x_1.iter().map(|x| (*x).into()).collect();
```
Earlier, the setup functions returned `(Vec<G1Projective>,Vec<G2Projective>)` so we didn't have to write conversions like this into the setup functions.

arifiorino commented on src/basic_block/cq.rs:
```diff
@@ -35,40 +54,52 @@ impl BasicBlock for CQBasicBlock {
-    setup.extend(L_i_0_x_1);
-    return (setup, vec![T_x_2]);
+    let L_i_x_1 = L_i_x_1.iter().map(|x| (*x).into()).collect();
+    let L_i_0_x_1 = L_i_0_x_1.iter().map(|x| (*x).into()).collect();
```
Earlier, the setup functions returned `(Vec<G1Projective>,Vec<G2Projective>)` so we didn't have to write conversions like this into the setup functions.

arifiorino commented on src/basic_block.rs:
```diff
@@ -80,24 +89,66 @@ impl DataEnc {
+    SetupType::None
+  }
+
+  fn run(&self, _weights: &ArrayD<Fr>, _inputs: &Vec<&ArrayD<Fr>>) -> Vec<ArrayD<Fr>> {
```
Why change this to `weights`? It could either be the weights, or it could be tables.

liliatangxy replied:
For a basic block without weights, it will be an empty array

arifiorino commented on src/basic_block.rs:
```diff
@@ -80,24 +89,66 @@ impl DataEnc {
+    Err(format!("{:?} does not use weights.", self.block_type()))
   }
+
+  fn setup(&self, _srs: &SRS, _model: &ArrayD<Fr>) -> SetupType {
```
I thought we weren't using `model` anymore.

arifiorino commented on src/basic_block.rs:
```diff
@@ -110,3 +161,22 @@ pub trait BasicBlock {
+/// the usage of BasicBlocks in a graph. This prevents redundant setups for
+/// the same BasicBlock (e.g., we use CQ2 for ReLU and Log, but we don't
+/// want to setup CQ2 twice for ReLU)
+impl std::hash::Hash for dyn BasicBlock {
```
We could just define a `HashMap` with strings as keys instead, and run the `Debug` trait to generate the keys.

arifiorino reviewed (changes requested):
I left several detailed comments about this PR. In addition to these, this PR is not able to load an onnx file at all. There are several subtleties in loading an onnx file, and none of these are addressed here.

liliatangxy committed:
rebase

liliatangxy committed:
fix default for block_type

liliatangxy committed:
fix merge
